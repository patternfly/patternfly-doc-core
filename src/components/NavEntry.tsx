import { NavItem } from '@patternfly/react-core'

export interface TextContentEntry {
  id: string
  data: {
    id: string
    section: string
  }
  collection: string
}

interface NavEntryProps {
  entry: TextContentEntry
  isActive: boolean
}

export const NavEntry = ({ entry, isActive }: NavEntryProps) => {
  const { id } = entry
  const { id: entryTitle, section } = entry.data

  return (
    <NavItem itemId={id} to={`/${section}/${id}`} isActive={isActive} id={`nav-entry-${id}`}>
      {entryTitle}
    </NavItem>
  )
}
