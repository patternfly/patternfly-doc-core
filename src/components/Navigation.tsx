import { useState, type ReactNode } from "react";
import {
  Nav,
  NavList,
  NavItem,
  PageSidebar,
  PageSidebarBody,
} from "@patternfly/react-core";
import { useStore } from "@nanostores/react";
import { isNavOpen } from "../stores/navStore";

interface NavOnSelectProps {
  groupId: number | string;
  itemId: number | string;
  to: string;
}

interface NavEntry {
  id: string;
  data: {
    title: string;
  };
  collection: string;
}

interface NavigationProps {
  navEntries: NavEntry[];
}

export const Navigation: React.FunctionComponent<NavigationProps> = ({
  navEntries,
}: NavigationProps) => {
  const [activeItem, setActiveItem] = useState("");

  const onNavSelect = (
    _event: React.FormEvent<HTMLInputElement>,
    selectedItem: NavOnSelectProps
  ) => {
    typeof selectedItem.itemId === "string" &&
      setActiveItem(selectedItem.itemId);
  };

  const $isNavOpen = useStore(isNavOpen);

  const sortedNavEntries = navEntries.sort((a, b) =>
    a.data.title.localeCompare(b.data.title)
  );

  const navItems = sortedNavEntries.map((entry) => (
    <NavItem
      key={entry.id}
      itemId={entry.id}
      isActive={activeItem === entry.id}
      to={`/${entry.collection}/${entry.id}`}
    >
      {entry.data.title}
    </NavItem>
  ));

  return (
    <PageSidebar isSidebarOpen={$isNavOpen}>
      <PageSidebarBody>
        <Nav onSelect={onNavSelect}>
          <NavList>{navItems}</NavList>
        </Nav>
      </PageSidebarBody>
    </PageSidebar>
  );
};
