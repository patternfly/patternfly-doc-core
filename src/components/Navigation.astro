---
import { getCollection } from 'astro:content'

import { Navigation as ReactNav } from './Navigation.tsx'
import { TextContentEntry } from './NavEntry.tsx'

import { content } from '../content'

import { config } from '../pf-docs.config.mjs'

const collections = await Promise.all(
  content.map(
    async (entry) => await getCollection(entry.name as 'textContent'),
  ),
)

const navDataRaw = collections.flat();
const uniqueSections = new Set(navDataRaw.map((entry) => entry.data.section));
const navData: Record<string, TextContentEntry[]> = {};

const [orderedSections, unorderedSections] = Array.from(uniqueSections).reduce(
    (acc, section) => {
      if (!config.navSectionOrder) {
        acc[1].push(section)
        return acc
      }

      const index = config.navSectionOrder.indexOf(section)
      if (index > -1) {
        acc[0][index] = section
      } else {
        acc[1].push(section)
      }
      return acc
    },
    [[], []] as [string[], string[]],
  )

const sortedSections = [...orderedSections, ...unorderedSections.sort()]
sortedSections.map((section) => {
  const entries = navDataRaw
    .filter((entry) => entry.data.section === section)
    .map(entry => ({ id: entry.id, data: { id: entry.data.id, section, sortValue: entry.data.sortValue }} as TextContentEntry))

  let uniqueEntries = entries
  if (section === 'components' || section === 'layouts') {
    // only display unique entry.data.id in the nav list if the section is components
    uniqueEntries = [
      ...entries
        .reduce((map, entry) => {
          if (!map.has(entry.data.id)) {
            map.set(entry.data.id, entry)
          }
          return map
        }, new Map())
        .values(),
    ]
  }

    // Sort alphabetically, unless a sort value is specified in the frontmatter
  const sortedUniqueEntries = uniqueEntries.sort((a, b) => {
    if (a.data.sortValue || b.data.sortValue) {
      const aSortOrder = a.data.sortValue || 50
      const bSortOrder = b.data.sortValue || 50

      return aSortOrder - bSortOrder
    }
    return a.data.id.localeCompare(b.data.id)
  })

  navData[section] = sortedUniqueEntries;
})

---

<ReactNav client:only="react" navData={navData} transition:animate="fade" />
