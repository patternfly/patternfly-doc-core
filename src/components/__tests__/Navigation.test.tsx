import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { Navigation } from '../Navigation'
import { TextContentEntry } from '../NavEntry'

const mockEntries: TextContentEntry[] = [
  {
    id: 'entry1',
    data: { id: 'Entry1', section: 'section1' },
    collection: 'textContent',
  },
  {
    id: 'entry2',
    data: { id: 'Entry2', section: 'section1' },
    collection: 'textContent',
  },
  {
    id: 'entry3',
    data: { id: 'Entry3', section: 'section2' },
    collection: 'textContent',
  },
]

it('renders without crashing', () => {
  render(<Navigation navEntries={mockEntries} />)
  expect(screen.getByText('Section1')).toBeInTheDocument()
  expect(screen.getByText('Section2')).toBeInTheDocument()
})

it('renders the correct number of sections', () => {
  render(<Navigation navEntries={mockEntries} />)
  expect(screen.getAllByRole('listitem')).toHaveLength(2)
})

it('sets the active item based on the current pathname', () => {
  Object.defineProperty(window, 'location', {
    value: {
      pathname: '/section1/entry1',
    },
    writable: true,
  })

  render(<Navigation navEntries={mockEntries} />)

  const entryLink = screen.getByRole('link', { name: 'Entry1' })

  expect(entryLink).toHaveClass('pf-m-current')
})

it('updates the active item on selection', async () => {
  const user = userEvent.setup()

  render(<Navigation navEntries={mockEntries} />)

  const entryLink = screen.getByRole('link', { name: 'Entry2' })

  await user.click(entryLink)

  expect(entryLink).toHaveClass('pf-m-current')
})

it('matches snapshot', () => {
  const { asFragment } = render(<Navigation navEntries={mockEntries} />)
  expect(asFragment()).toMatchSnapshot()
})
