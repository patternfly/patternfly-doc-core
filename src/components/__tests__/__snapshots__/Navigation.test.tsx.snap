// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    aria-hidden="false"
    class="pf-v6-c-page__sidebar pf-m-expanded"
    id="page-sidebar"
  >
    <div
      class="pf-v6-c-page__sidebar-body"
    >
      <nav
        aria-label="Global"
        class="pf-v6-c-nav"
        data-ouia-component-id="OUIA-Generated-Nav-3"
        data-ouia-component-type="PF6/Nav"
        data-ouia-safe="true"
      >
        <ul
          class="pf-v6-c-nav__list"
          role="list"
        >
          <li
            class="pf-v6-c-nav__item pf-m-expanded pf-m-current"
            data-ouia-component-id="OUIA-Generated-NavExpandable-5"
            data-ouia-component-type="PF6/NavExpandable"
            data-ouia-safe="true"
          >
            <button
              aria-expanded="true"
              class="pf-v6-c-nav__link"
              id="nav-section-section1"
            >
              Section1
              <span
                class="pf-v6-c-nav__toggle"
              >
                <span
                  class="pf-v6-c-nav__toggle-icon"
                >
                  <svg
                    aria-hidden="true"
                    class="pf-v6-svg"
                    fill="currentColor"
                    height="1em"
                    role="img"
                    viewBox="0 0 256 512"
                    width="1em"
                  >
                    <path
                      d="M224.3 273l-136 136c-9.4 9.4-24.6 9.4-33.9 0l-22.6-22.6c-9.4-9.4-9.4-24.6 0-33.9l96.4-96.4-96.4-96.4c-9.4-9.4-9.4-24.6 0-33.9L54.3 103c9.4-9.4 24.6-9.4 33.9 0l136 136c9.5 9.4 9.5 24.6.1 34z"
                    />
                  </svg>
                </span>
              </span>
            </button>
            <section
              aria-labelledby="nav-section-section1"
              class="pf-v6-c-nav__subnav"
            >
              <ul
                class="pf-v6-c-nav__list"
                role="list"
              >
                <li
                  class="pf-v6-c-nav__item"
                  data-ouia-component-id="OUIA-Generated-NavItem-7"
                  data-ouia-component-type="PF6/NavItem"
                  data-ouia-safe="true"
                >
                  <a
                    aria-current="page"
                    class="pf-v6-c-nav__link pf-m-current"
                    href="/section1/entry1"
                    id="nav-entry-entry1"
                  >
                    <span
                      class="pf-v6-c-nav__link-text"
                    >
                      Entry1
                    </span>
                  </a>
                </li>
                <li
                  class="pf-v6-c-nav__item"
                  data-ouia-component-id="OUIA-Generated-NavItem-8"
                  data-ouia-component-type="PF6/NavItem"
                  data-ouia-safe="true"
                >
                  <a
                    class="pf-v6-c-nav__link"
                    href="/section1/entry2"
                    id="nav-entry-entry2"
                  >
                    <span
                      class="pf-v6-c-nav__link-text"
                    >
                      Entry2
                    </span>
                  </a>
                </li>
              </ul>
            </section>
          </li>
          <li
            class="pf-v6-c-nav__item"
            data-ouia-component-id="OUIA-Generated-NavExpandable-6"
            data-ouia-component-type="PF6/NavExpandable"
            data-ouia-safe="true"
          >
            <button
              aria-expanded="false"
              class="pf-v6-c-nav__link"
              id="nav-section-section2"
            >
              Section2
              <span
                class="pf-v6-c-nav__toggle"
              >
                <span
                  class="pf-v6-c-nav__toggle-icon"
                >
                  <svg
                    aria-hidden="true"
                    class="pf-v6-svg"
                    fill="currentColor"
                    height="1em"
                    role="img"
                    viewBox="0 0 256 512"
                    width="1em"
                  >
                    <path
                      d="M224.3 273l-136 136c-9.4 9.4-24.6 9.4-33.9 0l-22.6-22.6c-9.4-9.4-9.4-24.6 0-33.9l96.4-96.4-96.4-96.4c-9.4-9.4-9.4-24.6 0-33.9L54.3 103c9.4-9.4 24.6-9.4 33.9 0l136 136c9.5 9.4 9.5 24.6.1 34z"
                    />
                  </svg>
                </span>
              </span>
            </button>
            <section
              aria-labelledby="nav-section-section2"
              class="pf-v6-c-nav__subnav"
              hidden=""
            >
              <ul
                class="pf-v6-c-nav__list"
                role="list"
              >
                <li
                  class="pf-v6-c-nav__item"
                  data-ouia-component-id="OUIA-Generated-NavItem-9"
                  data-ouia-component-type="PF6/NavItem"
                  data-ouia-safe="true"
                >
                  <a
                    class="pf-v6-c-nav__link"
                    href="/section2/entry3"
                    id="nav-entry-entry3"
                  >
                    <span
                      class="pf-v6-c-nav__link-text"
                    >
                      Entry3
                    </span>
                  </a>
                </li>
              </ul>
            </section>
          </li>
        </ul>
      </nav>
    </div>
  </div>
</DocumentFragment>
`;
